web/src/compose_banner.ts

```diff
@@ -78,12 +81,21 @@ export function update_or_append_banner(
         append_compose_banner_to_banner_list(banner, $list_container);
     } else {
         $banner.replaceWith(banner);
+        visible_banners.delete($banner);
     }
+    visible_banners.add(banner);
 }
 
 export function clear_message_sent_banners(include_unmute_banner = true): void {
     for (const classname of Object.values(MESSAGE_SENT_CLASSNAMES)) {
-        $(`#compose_banners .${CSS.escape(classname)}`).remove();
+        const $banners = $(`#compose_banners .${CSS.escape(classname)}`);
```

Comment:
I thought I commented on this earlier, but maybe it was in a different PR? I don't think reaching into the DOM is the right approach here. Why not make better use of the `visible_banners` Set and have that be the source of truth for the state of banners in the compose box? I see things get added, deleted, and cleared from that set here, but no where does it seem like that set has any bearing on the banners that are actually being displayed? Am I missing something?

Reply:
https://github.com/zulip/zulip/blob/2a8ecd02cac3e2e06f733d0a710591fa97a66bf5/web/src/compose_validate.js#L132-L134

so far the `visible_banners` set has been used over here to check if the banner already exists or not . currently the set stores HTMLStrings instead of an object that consists of things like `classnames` which seems to be needed here as the function itself doesn't seem to accept a particular banner string to be deleted. so instead i ended up going through the elements that are getting selected one by one and clearing their strings from the set individually.

Reply:
as an alternative instead of removing each banner individually i could simply clear the set but that would potentially lead to bugs since the `visible_banners` is container for all types of banners ( not just warning banners ).

Reply:
So why not store that information as part of an entry in the set?

Reply:
the ` append_compose_banner_to_banner_list`  where i am adding the banners to the set function seems to have access to only the HTMLString and its container as the parameters. I will probably have to add more arguments to the function to get access to `banner type` and `stream name` which could lead to fundamental code changes throughout the compose banner area since this function seems to be used to add the banners at all places.